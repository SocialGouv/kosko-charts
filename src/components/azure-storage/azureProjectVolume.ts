import { assert } from "@sindresorhus/is";
import environments from "@socialgouv/kosko-charts/environments";
import { updateMetadata } from "@socialgouv/kosko-charts/utils/updateMetadata";
import { PersistentVolume, PersistentVolumeClaim } from "kubernetes-models/v1";

export function azureProjectVolume(
  name: string,
  { storage }: { storage: string }
): [PersistentVolumeClaim, PersistentVolume] {
  const ciEnv = environments(process.env);
  assert.nonEmptyObject(ciEnv.metadata.annotations);
  assert.nonEmptyObject(ciEnv.metadata.labels);

  const envName =
    ciEnv.metadata.annotations["app.gitlab.com/env"] ||
    ciEnv.metadata.labels.application;

  const metadata = {
    annotations: ciEnv.metadata.annotations,
    labels: ciEnv.metadata.labels,
    namespace: ciEnv.metadata.namespace,
  };
  const pv = `${envName}-${name}`;

  const persistentVolumeClaim = new PersistentVolumeClaim({
    metadata: {
      name,
    },
    spec: {
      accessModes: ["ReadWriteMany"],
      resources: {
        requests: {
          storage,
        },
      },
      selector: {
        matchLabels: {
          usage: pv,
        },
      },
      storageClassName: "",
    },
  });

  const persistentVolume = new PersistentVolume({
    metadata: {
      labels: {
        usage: pv,
      },
      name: pv,
    },
    spec: {
      accessModes: ["ReadWriteMany"],
      azureFile: {
        secretName: `azure-${ciEnv.metadata.labels.team}-volume`,
        secretNamespace: ciEnv.metadata.namespace.name,
        shareName: name,
      },
      capacity: {
        storage,
      },
      persistentVolumeReclaimPolicy: "Delete",
      storageClassName: "",
    },
  });

  updateMetadata(persistentVolumeClaim, metadata);
  updateMetadata(persistentVolume, metadata);
  return [persistentVolumeClaim, persistentVolume];
}
