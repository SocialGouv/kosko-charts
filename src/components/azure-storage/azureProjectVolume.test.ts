import { azureProjectVolume } from "./azureProjectVolume";

jest.mock("@socialgouv/kosko-charts/environments/gitlab", () => ({
  // eslint-disable-next-line @typescript-eslint/naming-convention
  __esModule: true,
  default: () => ({
    metadata: {
      annotations: {
        "app.gitlab.com/app": "socialgouv-sample",
        "app.gitlab.com/env": "my-test",
      },
      labels: {
        application: "sample",
        owner: "sample",
        team: "sample",
      },
      namespace: { name: "sample-42-my-test" },
    },
    projectName: "sample",
    shortSha: "abcdefg",
  }),
}));

beforeEach(() => {
  jest.resetModules();
});

test("should create a persistent volume claim and a persistent volume", () => {
  expect(azureProjectVolume("foo", { storage: "42o" })).toMatchSnapshot();
});
