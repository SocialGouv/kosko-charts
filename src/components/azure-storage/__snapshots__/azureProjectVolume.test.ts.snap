// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a persistent volume claim and a persistent volume 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "PersistentVolumeClaim",
    "metadata": Object {
      "annotations": Object {
        "app.gitlab.com/app": "socialgouv-sample",
        "app.gitlab.com/env": "my-test",
      },
      "labels": Object {
        "application": "sample",
        "owner": "sample",
        "team": "sample",
      },
      "name": "foo",
      "namespace": "sample-42-my-test",
    },
    "spec": Object {
      "accessModes": Array [
        "ReadWriteMany",
      ],
      "resources": Object {
        "requests": Object {
          "storage": "42o",
        },
      },
      "selector": Object {
        "matchLabels": Object {
          "usage": "my-test-foo",
        },
      },
      "storageClassName": "",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "PersistentVolume",
    "metadata": Object {
      "annotations": Object {
        "app.gitlab.com/app": "socialgouv-sample",
        "app.gitlab.com/env": "my-test",
      },
      "labels": Object {
        "application": "sample",
        "owner": "sample",
        "team": "sample",
        "usage": "my-test-foo",
      },
      "name": "my-test-foo",
      "namespace": "sample-42-my-test",
    },
    "spec": Object {
      "accessModes": Array [
        "ReadWriteMany",
      ],
      "azureFile": Object {
        "secretName": "azure-sample-volume",
        "secretNamespace": "sample-42-my-test",
        "shareName": "foo",
      },
      "capacity": Object {
        "storage": "42o",
      },
      "persistentVolumeReclaimPolicy": "Delete",
      "storageClassName": "",
    },
  },
]
`;
